edlib/src/edlib.tpp

```diff
@@ -1417,51 +1437,58 @@ static int obtainAlignmentHirschberg(
  * @return  Alphabet as a string of unique characters, where index of each character is its value in transformed
  *          sequences.
  */
-static string transformSequences(const char* const queryOriginal, const int queryLength,
-                                 const char* const targetOriginal, const int targetLength,
-                                 unsigned char** const queryTransformed,
-                                 unsigned char** const targetTransformed) {
+template <class Element, class AlphabetIdx>
+unordered_map<Element, AlphabetIdx> transformSequences(
+        const Element* const queryOriginal, const int queryLength,
+        const Element* const targetOriginal, const int targetLength,
+        AlphabetIdx** const queryTransformed,
+        AlphabetIdx** const targetTransformed) {
     // Alphabet is constructed from letters that are present in sequences.
     // Each letter is assigned an ordinal number, starting from 0 up to alphabetLength - 1,
     // and new query and target are created in which letters are replaced with their ordinal numbers.
     // This query and target are used in all the calculations later.
-    *queryTransformed = static_cast<unsigned char *>(malloc(sizeof(unsigned char) * queryLength));
-    *targetTransformed = static_cast<unsigned char *>(malloc(sizeof(unsigned char) * targetLength));
+    *queryTransformed = static_cast<AlphabetIdx *>(malloc(sizeof(AlphabetIdx) * queryLength));
+    *targetTransformed = static_cast<AlphabetIdx *>(malloc(sizeof(AlphabetIdx) * targetLength));
 
-    string alphabet = "";
+    unordered_map<Element, AlphabetIdx> elementToAlphabetIdx;
 
     // Alphabet information, it is constructed on fly while transforming sequences.
-    // letterIdx[c] is index of letter c in alphabet.
-    unsigned char letterIdx[MAX_UCHAR + 1];
-    bool inAlphabet[MAX_UCHAR + 1]; // inAlphabet[c] is true if c is in alphabet
-    for (int i = 0; i < MAX_UCHAR + 1; i++) inAlphabet[i] = false;
-
+    // elementToAlphabetIdx[c] is index of letter c in alphabet.
+    AlphabetIdx currentSize = 0 ;
     for (int i = 0; i < queryLength; i++) {
-        unsigned char c = static_cast<unsigned char>(queryOriginal[i]);
-        if (!inAlphabet[c]) {
-            inAlphabet[c] = true;
-            letterIdx[c] = static_cast<unsigned char>(alphabet.size());
-            alphabet += queryOriginal[i];
+        Element c = queryOriginal[i];
+        //get an iterator to the element, c, in the alphabet hash table
+        auto iter = elementToAlphabetIdx.find(c);
+        if (iter == elementToAlphabetIdx.end()) { // if the element is not inserted previously
+            elementToAlphabetIdx[c] = currentSize;
+            (*queryTransformed)[i] = currentSize;
+            currentSize++;
```

Comment:
One tricky thing that could happen here is that we get an overflow, if alphabet is larger than the type we are using for it!
We should detect that and fail in such case.

The best way to do this might be to define our custom AlphabetTooBig exception and throw it here -> it would in its message explain that larger type for alphabet should be used. Then, users can catch it if they want, or not.